### PR TITLE
connectd: fix accidental handling of old reconnections.

### PR DESCRIPTION
**Pull request by rustyrussell**
We had multiple reports of channels being unilaterally closed because
it seemed like the peer was sending old revocation numbers.

Turns out, it was actually old reestablish messages!  When we have a
reconnection, we would put the new connection aside, and tell lightningd
to close the current connection: when it did, we would restart
processing of the initial reconnection.

However, we could end up with *multiple* "reconnecting" connections,
while waiting for an existing connection to close.  Though the
connections were long gone, there could still be messages queued
(particularly the channel_reestablish message, which comes early on).

Eventually, a normal reconnection would cause us to process one of
these reconnecting connections, and channeld would see the (perhaps
very old!) messages, and get confused.

(I have a test which triggers this, but it also hangs the connect
 command, due to other issues we will fix in the next release...)

Fixes: #5240
Fixes: #5235

### CODE REVIEW

**Comment by niftynei** [comment]
ACK https://github.com/ElementsProject/lightning/pull/5256/commits/3fab422747facaad34fd8d32c02d75d2718d550d

**Review by vincenzopalazzo — APPROVED** [review]
ACK https://github.com/ElementsProject/lightning/pull/5256/commits/3fab422747facaad34fd8d32c02d75d2718d550d

This should fix also https://github.com/ElementsProject/lightning/issues/5235 just to free all the issue :)

**Comment by ZmnSCPxj** [comment]
Looks OK to merge, just a style nit: the checks in our repo look for `Changelog-whatever` tags in commit messages, but this specific PR puts the changelog entries directly into the `CHANGELOG.md` file.

File: connectd/connectd.c [hunk 1]
```diff
@@ -259,6 +252,14 @@ static struct io_plan *peer_reconnected(struct io_conn *conn,
 
 	status_peer_debug(id, "reconnect");
 
+	/* If we have a previous reconnection, we replace it. */
+	pr = peer_reconnected_htable_get(&daemon->reconnected, id);
+	if (pr) {
+		status_broken("Reconnected AGAIN");
+		peer_reconnected_htable_del(&daemon->reconnected, pr);
```
Comment (ZmnSCPxj) [inline connectd/connectd.c:259]: If the previous reconnection already had the destructor installed, would it not delete this twice?  Or delete from `htable` is idempotent?

**Comment by rustyrussell** [comment]
Removed accidentally left BROKEN log, and added another assert removal to fix another report.

**Review by vincenzopalazzo — APPROVED** [review]
ACK https://github.com/ElementsProject/lightning/pull/5256/commits/b817e0a3ae5ec75f1f8dcf27f6d54d7a7f445e46

frustating I know, but I think this will fix also https://github.com/ElementsProject/lightning/issues/5254

**Comment by cdecker** [comment]
ACK b817e0a3ae5ec75f1f8dcf27f6d54d7a7f445e46

**Comment by svewa** [comment]
`bad reestablish revocation_number: 0 vs 3` in the changelog is maybe a bad example, as a revocation number of `0` usually indicates use of SCB?

File: CHANGELOG.md [hunk 2]
```diff
@@ -8,6 +8,18 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
 
+## [0.11.1] - 2022-05-13: Simon's Carefully Chosen Release Name II
+
+Single change which fixed a bug introduced in 0.11.0 which could cause
+unwanted unilateral closes (`bad reestablish revocation_number: 0 vs 3`)
+
+### Fixed
+
+ - connectd: make sure we don't keep stale reconnections around. ([#5256])
+ - connectd: fix assert which we could trigger. ([#5256])
```
Comment (vincenzopalazzo) [inline CHANGELOG.md:19]:
```suggestion
 - connectd: fix assert which we could trigger. ([#5254])
```

This should be https://github.com/ElementsProject/lightning/issues/5254 ?

Comment (rustyrussell) [reply CHANGELOG.md:19]: No, these refer to the PR, not the bug itself...

Comment (vincenzopalazzo) [reply CHANGELOG.md:19]: Ah damnt! Sorry just assuming wrong stuff

**Pull request merged**